File: php/tests/Unit/Domain/Catalogue/ValueObject/MoneyTest.php
<?php

declare(strict_types=1);

namespace Tests\Unit\Domain\Catalogue\ValueObject;

use Alxarafe\App\Domain\Catalogue\ValueObject\Money;
use InvalidArgumentException;
use PHPUnit\Framework\TestCase;

final class MoneyTest extends TestCase
{
    public function testCreateMoney(): void
    {
        $money = new Money(100.50, 'EUR');
        self::assertSame(100.50, $money->amount());
        self::assertSame('EUR', $money->currency());
    }

    public function testZeroAmount(): void
    {
        $money = new Money(0.0, 'USD');
        self::assertSame(0.0, $money->amount());
    }

    public function testNegativeAmount(): void
    {
        $this->expectException(InvalidArgumentException::class);
        new Money(-1.0, 'EUR');
    }

    public function testInvalidCurrencyTooShort(): void
    {
        $this->expectException(InvalidArgumentException::class);
        new Money(10.0, 'EU');
    }

    public function testInvalidCurrencyTooLong(): void
    {
        $this->expectException(InvalidArgumentException::class);
        new Money(10.0, 'EURO');
    }

    public function testInvalidCurrencyLowercase(): void
    {
        $this->expectException(InvalidArgumentException::class);
        new Money(10.0, 'eur');
    }

    public function testEquals(): void
    {
        $a = new Money(100.0, 'EUR');
        $b = new Money(100.0, 'EUR');
        self::assertTrue($a->equals($b));
    }

    public function testEqualsDifferentAmount(): void
    {
        $a = new Money(100.0, 'EUR');
        $b = new Money(200.0, 'EUR');
        self::assertFalse($a->equals($b));
    }

    public function testEqualsDifferentCurrency(): void
    {
        $a = new Money(100.0, 'EUR');
        $b = new Money(100.0, 'USD');
        self::assertFalse($a->equals($b));
    }
}

File: java/src/test/java/com/alxarafe/domain/catalogue/valueobject/MoneyTest.java
package com.alxarafe.domain.catalogue.valueobject;

import com.alxarafe.app.domain.catalogue.valueobject.Money;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class MoneyTest {

    @Test
    void createMoney() {
        var money = new Money(100.50, "EUR");
        assertEquals(100.50, money.amount());
        assertEquals("EUR", money.currency());
    }

    @Test
    void zeroAmount() {
        var money = new Money(0.0, "USD");
        assertEquals(0.0, money.amount());
    }

    @Test
    void negativeAmount() {
        assertThrows(IllegalArgumentException.class, () -> new Money(-1.0, "EUR"));
    }

    @Test
    void nullCurrency() {
        assertThrows(NullPointerException.class, () -> new Money(10.0, null));
    }

    @Test
    void invalidCurrencyTooShort() {
        assertThrows(IllegalArgumentException.class, () -> new Money(10.0, "EU"));
    }

    @Test
    void invalidCurrencyTooLong() {
        assertThrows(IllegalArgumentException.class, () -> new Money(10.0, "EURO"));
    }

    @Test
    void invalidCurrencyLowercase() {
        assertThrows(IllegalArgumentException.class, () -> new Money(10.0, "eur"));
    }

    @Test
    void equals() {
        var a = new Money(100.0, "EUR");
        var b = new Money(100.0, "EUR");
        assertEquals(a, b);
    }

    @Test
    void equalsDifferentAmount() {
        var a = new Money(100.0, "EUR");
        var b = new Money(200.0, "EUR");
        assertNotEquals(a, b);
    }

    @Test
    void equalsDifferentCurrency() {
        var a = new Money(100.0, "EUR");
        var b = new Money(100.0, "USD");
        assertNotEquals(a, b);
    }
}
